src/gui/SampleLoader.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * SampleLoader.cpp - Static functions that open audio files
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleLoader.h"
+
+#include <QFileInfo>
+#include <QMessageBox>
+
+#include "ConfigManager.h"
+#include "FileDialog.h"
+#include "PathUtil.h"
+
+namespace lmms::gui {
+QString SampleLoader::openAudioFile(const QString& previousFile)
+{
+	auto openFileDialog = gui::FileDialog(nullptr, QObject::tr("Open audio file"));
+	auto dir = QString{};
+
+	if (!previousFile.isEmpty())
+	{
+		auto file = QString{previousFile};
+		if (QFileInfo{file}.isRelative())
+		{
+			file = ConfigManager::inst()->userSamplesDir() + file;
+			if (!QFileInfo{file}.exists()) { file = ConfigManager::inst()->factorySamplesDir() + previousFile; }
```

Comment:
If the file doesn't exist on line 46, `file` will be an empty string, and I'm wondering if the rest of the method would behave in an unintended way if that happened.

Reply:
Isn't it assigned on line 45?

Reply:
Sorry, yes. It could be non-empty but still be the path to a nonexistent file if `ConfigManager::inst()->factorySamplesDir() + previousFile` doesn't exist.

Reply:
> Sorry, yes. It could be non-empty but still be the path to a nonexistent file if ConfigManager::inst()->factorySamplesDir() + previousFile doesn't exist.

Well, if ``ConfigManager::inst()->userSamplesDir() + file;`` doesn't exist, and then ``ConfigManager::inst()->factorySamplesDir() + previousFile;`` doesn't exist, I'm not sure if there is an optimal fallback. I doubt these directories won't exist though, and as previously mentioned, this was mostly pulled from the old ``SampleBuffer`` but changed a bit, so functionality wise, these functions seem fine.